mindmeld/models/nn_utils/input_encoders.py

```diff
@@ -0,0 +1,937 @@
+
+    @classmethod
+    @abstractmethod
+    def load(cls, **kwargs):
+        msg = f"Subclass {cls.__name__} need to implement this method"
+        raise NotImplementedError(msg)
+
+    @abstractmethod
+    def batch_encode(self, examples, labels=None, **kwargs) -> Dict[str, Any]:
+        msg = f"Subclass {self.name} need to implement this method"
+        raise NotImplementedError(msg)
+
+    # common `get` methods
+
+    def get_num_tokens(self):
+        if not hasattr(self, "token2idx"):
+            return None
+        return len(self.token2idx)
+
+    def get_emb_dim(self):
+        return self.emb_dim
+
+    def get_pad_token_idx(self):
+        if not hasattr(self, "pad_token_idx"):
+            return None
+        return self.pad_token_idx
+
+    def get_embedding_weights(self):
+        if not hasattr(self, "token2emb") or not self.token2emb:
+            msg = f"Encoder instance ({self.name}) does not contain any token-to-embeddings mapping"
+            logger.info(msg)
+            return None
+        embedding_weights = {}
+        for token, idx in self.token2idx.items():
+            if token in self.token2emb:
+                embedding_weights[idx] = self.token2emb[token]
+        return embedding_weights
+
+    @staticmethod
+    def get_pad_label_idx():
+        return LABEL_PAD_TOKEN_IDX
+
+    def _reformat_and_validate(self, examples, labels=None):
+
+        # reformatting
+        if isinstance(examples, str):
+            examples = [examples]
+            if labels:
+                labels = [labels]
+
+        # validation for labels size
+        if labels:
+            if len(examples) != len(labels):
+                msg = f"Number of 'labels' ({len(labels)}) must be same as number of 'examples' " \
+                      f"({len(examples)}) when passing labels to {self.name}.batch_encode()"
+                raise AssertionError(msg)
+
+        return examples, labels
+
+
+# base encoders
+
+
+class WordEmbeddingsBasedEncoder(AbstractEncoder):
+    """
+    A base class for joint tokenization-encoding with the vocab derived from the training data and
+    using word-level embedders if required.
+
+    TOKENIZATION SUPPORT:
+    This class supports three kinds of tokenizations: (1) tokenization at white space,
+    (2) character tokenizations
+
+    ENCODING SUPPORT
+    This class supports encoding text into 0-based ids for an embedding lookup. When using the
+    padding_length, at times, the labels might have to be trimmed as well, which is to be supported
+    in the .batch_encode() method of the derived class.
+    """
+
+    ALLOWED_TOKENIZERS = [None, "whitespace-tokenizer", "char-tokenizer"]
+    ALLOWED_EMBEDDER_TYPES = [None, "glove"]
+    BASIC_SPECIAL_VOCAB_DICT = {
+        "pad_token": "<PAD>",
+        "unk_token": "<UNK>",
+        "start_token": "<START>",
+        "end_token": "<END>",
+    }
+    DEFAULT_PARAMS = {
+        "emb_dim": 300,
+    }
+
+    def fit(
```

Comment:
while this is not an ML model, encoders are might sometimes need training like in case of sentence pice or BPE. training here means creating a stateful object which splits inputs based on some algo that is learnt with the trained data.